### PR TITLE
Implement sorting of package entries

### PR DESCRIPTION
**Pull request by ankush981**
Hey, I think this works now as expected. Hopefully. 😅 I took the liberty of changing the underlying flow a bit, resulting in major changes in the existing code. I'm not a seasoned Python dev, so please at east to criticize and suggest better ways. 😄

### CODE REVIEW

File: pyp/pyp.py [hunk 1]
```diff
@@ -20,22 +24,24 @@ def pip_freeze(package_names):
...
-                    else:
-                        f.write(pkg + '\n')
-            except FileNotFoundError:
-                open('requirements.txt', 'w').close()
-            else:
-                break
-
+    existing_entries = get_existing_entries()
+    new_entries = packages.split('\n')[:-1] # Skip the '' at the end
+    combined_entries = existing_entries + new_entries
```
Comment (zachtylr21) [inline pyp/pyp.py:39]: I noticed a small bug with this line. If you `pyp install` a package twice, the package will get written twice to the `requirements.txt`. Ideally, `pyp install` should be idempotent. I think an efficient way to fix this would be to import `OrderedDict` from `collections` and do
```suggestion
    combined_entries = list(OrderedDict.fromkeys(existing_entries + new_entries))
```
The reason I suggest using `OrderedDict` is because it removes duplicates while preserving the order of `existing_entries` which means the sorting algorithm doesn't have to do as much.

File: tests/test_all.py [hunk 2]
```diff
@@ -1,10 +1,16 @@
...
 def test_pip_freeze_search_patterns(mock_check_output):
     pip_freeze(['plain', 'with-hyphen', 'with_underscore', 'this-has_both'])
     mock_check_output.assert_called_once()
     mock_check_output.call_args[0][0] == "pip freeze | grep -i 'plain==\\|with[-_]hyphen==\\|with[-_]underscore==\\|this[-_]has[-_]both=='"
+
+def test_package_name_sorting_handles_irregular_names():
+    packages = ['Django-7.5', 'abc-33', 'rANndom_44', 'pip']
+    expected_output = ['abc-33', 'django-7.5', 'pip', 'ranndom_44']
+    sorted_packages = sort_entries(packages)
+    ''.join(sorted_packages) == ''.join(expected_output)
```
Comment (zachtylr21) [inline tests/test_all.py:16]: Would you mind adding a newline to the end of this file

**Review by zachtylr21 — CHANGES_REQUESTED** [review]
Thanks for helping out with this new feature! I commented on a few things I'd like to change, but overall, it looks good! I also noticed some extra spaces at the end of a few lines. If you could find and remove those that would be great.

**Comment by ankush981** [comment]
Hey, thanks for the encouragement and the suggestions. Honestly, I've learned a lot already through this project. 😄 About the newlines and all, I actually wanted to run the code through a formatter like Black, but then I saw you're using single quotes, which Black absolutely hates, so overall, we might have more clashes than resolutions. 😆

File: pyp/pyp.py [hunk 3]
```diff
@@ -20,22 +24,24 @@ def pip_freeze(package_names):
     ).decode('ascii')
     return output
 
+def get_existing_entries():
+    entries = []
+    try:    
+        entries = [entry.rstrip('\n') for entry in open(REQUIREMENTS_FILE_NAME)]
```
[the remark this answers is not in the document]
Comment (ankush981) [reply pyp/pyp.py:30]: Yes, sure. Makes sense!

File: pyp/pyp.py [hunk 4]
```diff
@@ -20,22 +24,24 @@ def pip_freeze(package_names):
...
-            else:
-                break
-
+    existing_entries = get_existing_entries()
+    new_entries = packages.split('\n')[:-1] # Skip the '' at the end
+    combined_entries = existing_entries + new_entries
+    sorted_entries = sort_entries(combined_entries)
+
+    file_contents = '\n'.join(sorted_entries) + '\n' # For the last entry
+    with open(REQUIREMENTS_FILE_NAME, 'w') as file: # Not sure about try/catch here
```
[the remark this answers is not in the document]
Comment (ankush981) [reply pyp/pyp.py:43]: Yeah, that's pretty much what I was thinking. 😄 There's also this edge case that what happens if the attempt to create the file fails, but I think now we're worrying too much.

File: tests/test_all.py (hunk 2 above)
Comment (ankush981) [reply tests/test_all.py:16]: Okay. I'm wondering why we're doing this, though? Python convention or conforming to the Unix file structure? Or maybe something else?

File: pyp/pyp.py (hunk 1 above)
Comment (ankush981) [reply pyp/pyp.py]: Hey, that's a very nice catch! The approach suggested was a bit of a mind-bender for me, but I eventually managed to understand it. 😅

**Comment by ankush981** [comment]
Hey, I've changed the code as per your suggestions. I wasn't able to get the tests to run at all, though I was able to do it easily last time. Maybe I've forgotten how I did it. 😅 Either way, can you please provide the instructions for running the tests? And while we're on that topic, are there are instructions for developers? If not, I'm happy to create a page dedicated to that.

File: pyp/pyp.py (hunk 1 above)
Comment (zachtylr21) [reply pyp/pyp.py]: Glad you figured it out!

File: tests/test_all.py (hunk 2 above)
Comment (zachtylr21) [reply tests/test_all.py]: I think having a newline at the end is a pep8 standard, and I've heard it can cause problems with running the script in some environments. Most likely it's fine either way, but I like to err on the safe side.

**Review by zachtylr21 — APPROVED** [review]
Thanks for making those changes! Good to hear that you've learned something 😄 

I think using an auto formatter is a great idea. I usually use autopep8 in other projects. I've never used black, but I've heard about it. I've heard yapf is another popular one. Since I don't have one officially picked out yet, I'll approve this PR and we can run a formatter after I've done some research and got one set up.

About the tests, the command is `pytest`. That should find the tests in the tests directory and run all of them.

I haven't written instructions for developers yet. What did you have in mind?

**Comment by ankush981** [comment]
Woooohooo! A non-trivial, successful contribution! 😄😄 By the way, I'm not among the ones temporarily motivated by Hacktoberfest fever; I always wanted to contribute but only now the conditions are right. So, if there are other nice projects, please let me know!

About the tests, I remember managing to run them earlier, but now `pytest` throws this ugly error:
```
int: make sure your test modules/packages have valid Python names.
Traceback:
../../../../.pyenv/versions/3.7.9/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_all.py:3: in <module>
    from pyp.pyp import pip_freeze, sort_entries
E   ModuleNotFoundError: No module named 'pyp'
```
About formatter, yes, a formatter will be very nice. Maybe even run it in a pre-commit hook or something.

**Comment by ankush981** [comment]
And as far as dev instructions go, I was thinking that having short, clear-cut steps will help. I mean, the person wishing to contribute may not know about virtual environments, or may create one in a new folder and send it along with the pull request. Or create a Python2 environment by mistake and wonder what's wrong. 

Then a couple of things from my experience: When I revisited this code to make the changes, it was just a week past, and I realized I had no idea how to run the project! I ended up doing `pip install pyp` followed by `pyp install` and wondering why the duplicates are still appearing! And then, later, as I mentioned, I couldn't figure out how to run the tests.

Finally, people may forget to clean up `requirements.txt` (since the project only needs `pytest`).

So, my suggestion is that having a list of 7-8 such instructions that people can just blindly follow can save a lot of minutes, if not hours, of pain. It also is a trivial effort on our part. 😄

**Comment by zachtylr21** [comment]
@ankush981 hmm... I'll have to look into that pytest error.

I like your ideas for instructions. I think we should create an issue where we can list out everything that should be included. I'll go ahead and create that issue now with the things you mentioned here. Feel free to add to it if you want.
